networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/compression/Compressor.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.eth2.compression;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import org.apache.tuweni.bytes.Bytes;
+
+public interface Compressor {
+
+  /**
+   * Returns the compressed data
+   *
+   * @param data The data to compress
+   * @return The compressed data
+   */
+  Bytes compress(final Bytes data);
+
+  /**
+   * Returns the uncompressed data.
+   *
+   * @param data The data to uncompress.
+   * @return The uncompressed data.
+   */
+  default Bytes uncompress(final Bytes data) throws CompressionException {
+    try (final InputStream byteStream = new ByteArrayInputStream(data.toArrayUnsafe())) {
+      // Read everything
+      return uncompress(byteStream, Integer.MAX_VALUE);
+    } catch (IOException e) {
+      throw new RuntimeException(
+          "Unexpected error encountered while preparing to uncompress bytes", e);
+    }
+  }
+
+  /**
+   * Uncompress up to {@code maxBytes} bytes. May return fewer bytes if the end of stream is
+   * detected or an error occurs.
+   *
+   * @param input The underlying {@link InputStream} to read from.
+   * @param maxBytes The maximum number of uncompressed bytes to produce
+   * @return The uncompressed bytes read from the underlying inputStream {@code input} stream.
+   */
+  Bytes uncompress(final InputStream input, final int maxBytes) throws CompressionException;
+
+  /**
+   * Returns a maximum estimate of the size of a compressed payload given the uncompressed payload
+   * size.
+   *
+   * @param uncompressedLength The size of an uncompressed payload.
+   * @return The maximum size of a payload of size {@code uncompressedLength} after compression.
+   */
+  int getMaxCompressedLength(final int uncompressedLength);
```

Comment:
I'm not 100% sure I've understood exactly where the issue enforcing this check is so these comments may be off the mark but...

I think the main security concern is that we can't be made to read massive amounts of data from the stream.  The total number of uncompressed bytes we could read is limited but there's a risk that we could be fed a series of effectively empty snappy frames so that we never reach that limit. Not sure if that's actually possible with snappy or if it would be rejected as invalid compressed data.

One option we could have is to limit the underlying input stream.  ie we have an `InputStream` that is the source of compressed data - we could wrap it in a custom `InputStream` that mostly just delegates but keeps track of the number of bytes read and throws an exception if we wind up reading more than the expected limit.

Reply:
Yeah, you're right.  I was hoping snappy was smarter about what constitutes a valid sequence of frames, but it will happily decode a sequence of identical frames each with a 1-byte payload: https://github.com/PegaSysEng/teku/pull/1670/commits/8eb87800d7f0267a055ee36e8a4687eeafcba857

So yeah, we'll have to have some kind of wrapped InputStream that tracks the number of bytes actually read.

Reply:
Turns out guava has a util for created a limited input stream.  It's marked `@Beta`: https://guava.dev/releases/19.0/api/docs/com/google/common/annotations/Beta.html.  But I think that's fine for us?  If not, I started on an implementation that we could swap in.

Reply:
I'm fine with using the Guava version.